lib/Basics/SizeLimitedString.h

```diff
@@ -0,0 +1,160 @@
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Jan Steemann
+////////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include <cstring>
+#include <string_view>
+#include <type_traits>
+
+#include "Basics/StringUtils.h"
+#include "Basics/debugging.h"
+
+namespace arangodb {
+
+// small utility class providing a fixed-size string.
+// the maximum length of the string is provided at compile time.
+// data can be added to the string until the fixed capacity is exceeded.
+// from that point on, all further append operations will do nothing
+// (i.e. not append to the buffer or overrun it), nor throw an exception.
+// all operations of this class are noexcept.
+template<size_t N>
+class SizeLimitedString {
+  static_assert(N > 0);
+
+ public:
+  SizeLimitedString(SizeLimitedString const&) = delete;
+  SizeLimitedString& operator=(SizeLimitedString const&) = delete;
+
+  SizeLimitedString() noexcept { clear(); }
+
+  std::string_view view() const noexcept {
+    TRI_ASSERT(_offset <= N);
+    return std::string_view(&_buffer[0], _offset);
+  }
+
+  void clear() noexcept {
+    memset(&_buffer[0], 0, sizeof(_buffer));
+    _offset = 0;
+    _full = false;
+  }
+
+  bool empty() const noexcept { return _offset == 0; }
+
+  size_t capacity() const noexcept { return N; }
+
+  size_t size() const noexcept {
+    TRI_ASSERT(N >= _offset);
+    return _offset;
+  }
+
+  void push_back(char c) noexcept {
+    if (_full) {
+      return;
+    }
+    _buffer[_offset] = c;
+    ++_offset;
+    recalculateState();
+  }
+
+  SizeLimitedString& append(std::string_view data) noexcept {
+    return append(data.data(), data.size());
+  }
+
+  SizeLimitedString& append(char const* data) noexcept {
+    return append(data, strlen(data));
+  }
+
+  SizeLimitedString& append(char const* data, size_t length) noexcept {
+    if (!_full) {
+      // append as much as is possible (potentially only a prefix)
+      length = std::min(length, remaining());
+      memcpy(_buffer + _offset, data, length);
+      _offset += length;
+      recalculateState();
+    }
+    return *this;
+  }
+
+  // append a uint64_t value as a string. will only append the value if
+  // there is enough capacity in the buffer for the maximum possible
+  // uint64_t value (which is 21 bytes long).
+  SizeLimitedString& appendUInt64(uint64_t value) noexcept {
+    if (!_full) {
+      if (basics::maxUInt64StringSize > remaining()) {
+        _full = true;
+      } else {
+        _offset += basics::StringUtils::itoa(value, _buffer + _offset);
+        recalculateState();
+      }
+    }
+    return *this;
+  }
+
+  // append a hex-encoded value.
+  // will append up to as many bytes as there is remaining space in the buffer.
+  template<typename T>
+  SizeLimitedString& appendHexValue(T value, bool stripLeadingZeros) noexcept {
```

Comment:
On a big endian platform this will write out the hex value in the wrong order. Should we bother?